File: src/main/java/cubesrv/ThreadPoolHelper.java
package cubesrv;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class ThreadPoolHelper {
    public static final int THREAD_COUNT = Runtime.getRuntime().availableProcessors();

    @FunctionalInterface
    public interface IntTask {
        void run(int threadNo) throws Exception;
    }

    public static void runInThreadPool(IntTask fn) {
        ExecutorService pool = Executors.newFixedThreadPool(THREAD_COUNT);
        List<Future<?>> futures = new ArrayList<>();
        for (int i = 0; i < THREAD_COUNT; i++) {
            final int t = i;
            futures.add(pool.submit(() -> { fn.run(t); return null; }));
        }
        for (Future<?> f : futures) {
            try {
                f.get();
            } catch (ExecutionException e) {
                throw new RuntimeException(e.getCause());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        }
        pool.shutdown();
    }
}

File: src/main/kotlin/cubesrv/ThreadPoolHelper.kt
package cubesrv

import java.lang.Runtime
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.launch

val THREAD_COUNT : Int = Runtime.getRuntime().availableProcessors()

fun runInThreadPool(fn : suspend (threadNo : Int) -> Unit) = runBlocking {
    for(threadNo in 0 ..< THREAD_COUNT) {
        launch(Dispatchers.Default) {
            fn(threadNo)
        }
    }
}
